### PR TITLE
C++: Remove abstract keyword from `BuiltInOperation`

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR removes the abstract keyword from the `BuiltInOperation` class. Here's my TODO list before I can move this out of draft mode:

- [x] Generate new stats file
- [x] Prepare internal PR (https://git.semmle.com/Semmle/code/pull/36883)

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/exprs/BuiltInOperations.qll [hunk 1]
```diff
@@ -4,15 +4,15 @@ import semmle.code.cpp.exprs.Expr
...
 
 /**
  * A C/C++ built-in operation that is used to support functions with variable numbers of arguments.
  * This includes `va_start`, `va_end`, `va_copy`, and `va_arg`.
  */
-class VarArgsExpr extends BuiltInOperation {
+class VarArgsExpr extends BuiltInOperation, @var_args_expr {
   VarArgsExpr() {
     this instanceof BuiltInVarArgsStart
     or
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/exprs/BuiltInOperations.qll:18]: If things are working properly, you can now remove this charpred. (Since it doesn't affect the dbscheme, you don't have to recreate the internal PR if you add a commit to this PR with such cosmetic changes.)

File: cpp/ql/src/semmle/code/cpp/exprs/BuiltInOperations.qll [hunk 2]
```diff
@@ -4,15 +4,15 @@ import semmle.code.cpp.exprs.Expr
  * A C/C++ builtin operation. This is the root QL class encompassing
  * built-in functionality.
  */
-abstract class BuiltInOperation extends Expr {
+class BuiltInOperation extends Expr, @buildin_op {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/exprs/BuiltInOperations.qll:7]:
```suggestion
class BuiltInOperation extends Expr, @builtin_op {
```
(plus the corresponding rename in the dbscheme)

**Comment by jbj** [comment]
Does the @github/codeql-c-extractor team have an opinion about this PR? Or know of any pitfalls around enumerating all builtins?

**Comment by matt-gretton-dann** [comment]
TL;DR: I defer to @igfoo & @nickrolfe who have longer history - but I think some improved docs are needed.

(Out of interest: What is the benefit to the C/C++ Analysis team for doing this change?)

My concern with the above is it is not a complete list of builtins that a compiler supports - rather it is the list of built-ins the extractor treats specially (for some reason).

The compiler has a much longer list of builtins (and I don't think they're all enumerated in one place for any particular compiler).  But for GCC see https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#Other-Builtins as a starting point.

The comment for `BuiltinOperation` isn't clear to me:

```ql
/**
 * A C/C++ builtin operation. This is the root QL class encompassing
 * built-in functionality.
 */
```

What is a C/C++ builtin operation?  Do we mean "C/C++ feature implemented by a __builtin_* function call (in GCC speak) that can't just be treated as a normal function or that a user didn't write as a function call"?

I think improving the docs and clarifying what we mean here would be a good thing, and therefore helping to decide whether this is a good change.

I defer to @igfoo and @nickrolfe as to whether there are other issues here though.

**Comment by MathiasVP** [comment]
> (Out of interest: What is the benefit to the C/C++ Analysis team for doing this change?)

Users extending abstract classes have caused cached stages to be recomputed (@jbj fixed one of these recently in https://github.com/github/codeql/pull/3319), so we tend to avoid them in user facing code. See https://jira.semmle.com/browse/CPP-458.

**Comment by igfoo** [comment]
Looks fine to me too.

The set of types in `builtintype` is the set that currently extend the abstract type. I agree that we should look at better documentation and coverage, but I think that that is orthogonal to this change.
